src/Features/Core/Portable/CodeRefactorings/AddAwait/AbstractAddAwaitCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,122 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.AddAwait
+{
+    /// <summary>
+    /// Refactor:
+    ///     var x = GetAsync();
+    ///
+    /// Into:
+    ///     var x = await GetAsync();
+    ///
+    /// Or:
+    ///     var x = await GetAsync().ConfigureAwait(false);
+    /// </summary>
+    internal abstract class AbstractAddAwaitCodeRefactoringProvider<TExpressionSyntax, TInvocationExpressionSyntax> : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+        where TInvocationExpressionSyntax : TExpressionSyntax
+    {
+        protected abstract string GetTitle();
+        protected abstract string GetTitleWithConfigureAwait();
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var textSpan = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (!textSpan.IsEmpty)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindTokenOnLeftOfPosition(textSpan.Start);
+
+            var model = await document.GetSemanticModelAsync(cancellationToken);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var awaitable = GetAwaitableExpression(textSpan, token, model, syntaxFacts, cancellationToken);
+            if (awaitable == null)
+            {
+                return;
+            }
+
+            if (awaitable.OverlapsHiddenPosition(cancellationToken))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    GetTitle(),
+                    c => AddAwaitAsync(document, awaitable, withConfigureAwait: false, c)));
+
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    GetTitleWithConfigureAwait(),
+                    c => AddAwaitAsync(document, awaitable, withConfigureAwait: true, c)));
+        }
+
+        private TExpressionSyntax GetAwaitableExpression(TextSpan textSpan, SyntaxToken token, SemanticModel model, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken)
+        {
+            var invocation = token.GetAncestor<TInvocationExpressionSyntax>();
+            if (invocation is null)
+            {
+                return null;
+            }
+
+            if (syntaxFacts.IsExpressionOfAwaitExpression(invocation))
```

Comment:
i would just do: `if syntaxFacts.IsAwaitExpression(invocation.parent)`

Reply:
We already have a specialized method for this. I feel strongly m'eh ;-)

Reply:
Oh, never mind then.  That's fine :)  I thought you added this yourself.